multipledispatch/dispatcher.py

```diff
@@ -281,6 +284,31 @@ def source(self, *args, **kwargs):
         print(self._source(*args))
 
 
+class _UnresolvedDispatcher(Dispatcher):
+    __slots__ = ()
+
+    @property
+    def __doc__(self):
+        # trick python into giving us our super class's dynamic __doc__
+        return super(_UnresolvedDispatcher, self).__doc__
+
+    @property
+    def ordering(self):
+        self.__class__ = self._resolved_type
+        self.reorder()
+        return self.ordering
+
+    @ordering.setter
+    def ordering(self, value):
+        self.__class__ = self._resolved_type
+        self.reorder()
+        self.ordering = value
```

Comment:
This looks like fiendish devilry to me.  I wonder if there is a less magical way to achieve the same result without significant cost. 

For example, I believe that try-except blocks are quite cheap if there is no exception.  I wonder if something like the following would work well enough:

```python
def add(...):
    del self._ordering
    ...

@property
def ordering(self):
    try:
        return self._ordering
    except AttributeError:
        self.reorder()
        return self._ordering
```

Reply:
try/except is actually somewhat expensive because it needs to setup a block and unwind it. In blaze we have many millions of calls to multiply dispatched functions so even small overhead gets magnified. The goal of this was to be have no overhead once the dispatcher is resolved.

Reply:
We don't touch ordering that often though (caching should catch most calls), and when we do it's well nested within several other try-except blocks and function calls.

Reply:
Here is a (very untrustworthy) microbenchmark

```python
In [1]: %%timeit
   ...: try:
   ...:     1 + 1
   ...: except:
   ...:     pass
   ...: 
The slowest run took 96.99 times longer than the fastest. This could mean that an intermediate result is being cached.
100000000 loops, best of 3: 18.6 ns per loop

In [2]: %timeit 1 + 1
100000000 loops, best of 3: 10.8 ns per loop
```

Compare that to function calls

```python
In [3]: def f():
   ...:     pass
   ...: 

In [4]: %timeit f()
The slowest run took 17.84 times longer than the fastest. This could mean that an intermediate result is being cached.
10000000 loops, best of 3: 69.9 ns per loop
```

It's unclear to me that adding try-except block would significantly harm performance.  Swapping out `__class__` would, I suspect, have unexpected consequences when interacting with all of the other libraries of the ecosystem.

Reply:
@llllllllll reminder that the ball is in your court here :)